### PR TITLE
Fix bundle auto-create by checking for IBaseReference.hasIdentifier() before validating URL-based references

### PR DESCRIPTION
**Pull request by lukedegruchy**
Closes https://github.com/hapifhir/hapi-fhir/issues/4736

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java [hunk 1]
```diff
@@ -4391,6 +4393,141 @@ public void testPlaceholderCreateTransactionRetry_NonMatchingPlaceholderReferenc
...
+		obsToCreate.getSubject().getIdentifier().setSystem("http://bar").setValue("http://something.something/b");
+
+		final Bundle bundle = new Bundle();
+		bundle.setType(BundleType.TRANSACTION);
+		bundle.addEntry().setResource(obsToCreate)
+			.getRequest()
+			.setMethod(HTTPVerb.POST)
+			.setUrl("Observation");
+		// execute
+		final Bundle actual = mySystemDao.transaction(mySrd, bundle);
```
Comment (samguntersmilecdr) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:4529]: question: No validation that only one was created is done after this execute, is this intentional?

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java [hunk 2]
```diff
@@ -518,7 +518,7 @@ private void extractResourceLinks(@Nonnull RequestPartitionId theRequestPartitio
 			return;
 		}
 
-		String baseUrl = nextId.getBaseUrl();
+		String baseUrl = (! nextReference.hasIdentifier()) ? nextId.getBaseUrl() : null;
```
Comment (samguntersmilecdr) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java:521]: nitpick (non-blocking): Smallest nitpick ever, the negation in this condition makes this harder to understand at a glance

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoCreatePlaceholdersR4Test.java [hunk 3]
```diff
@@ -399,6 +403,66 @@ public void testCreatePlaceholderWithMatchingInlineAndSubjectReferenceIdentifier
...
+		final IIdType placeholderPatId = placeholderPat.getIdElement();
+		assertEquals(createdObs.getSubject().getReference(), placeholderPatId.toUnqualifiedVersionless().getValueAsString());
+
+		/*
+		 * Should have a single identifier populated.
+		 */
+		assertEquals(1, placeholderPat.getIdentifier().size());
+		final List<Identifier> identifiers = placeholderPat.getIdentifier();
+		Identifier identifier = identifiers.get(0);
+		assertThat(identifier.getSystem(), is(equalTo("http://bar")));
```
Comment (AD1306) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoCreatePlaceholdersR4Test.java:438]: nitpick: Do you want maybe declare system and value to variables ?

**Review by AD1306 — APPROVED** [review]
thanks for fixing this !!

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoCreatePlaceholdersR4Test.java [hunk 4]
```diff
@@ -399,6 +403,66 @@ public void testCreatePlaceholderWithMatchingInlineAndSubjectReferenceIdentifier
...
+		final String identifierValue = "http://some-url-value/Organization/ID2";
+		final Patient patient = new Patient();
+		final Reference reference = new Reference()
+			.setReference(ResourceType.Organization.name() + "?identifier=" + FAKE_IDENTIFIER_SYSTEM + "|" + identifierValue)
+			.setIdentifier(new Identifier().setValue(identifierValue).setSystem(FAKE_IDENTIFIER_SYSTEM));
+		patient.setManagingOrganization(reference);
+
+		final IIdType patientId = myPatientDao.create(patient, mySrd).getId();
+
+		final Patient createdPatient = myPatientDao.read(patientId, new SystemRequestDetails());
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoCreatePlaceholdersR4Test.java]: The patient is pretty much empty so there's nothing to assert.

**Comment by vilaug** [comment]
I think this causes this issue: https://github.com/hapifhir/hapi-fhir/issues/5148, is it expected behavior?

**Pull request merged**